poly/src/polynomial/mod.rs

```diff
@@ -1,164 +1,93 @@
-    fn from(other: &'a SparsePolynomial<F>) -> Self {
-        SPolynomial(Cow::Borrowed(other))
-    }
-}
+    /// If `num_vars` is `None`, outputs a polynomial a univariate polynomial
+    /// of degree `d` where each coefficient is sampled uniformly at random.
+    ///
+    /// If `num_vars` is `Some(l)`, outputs an `l`-variate polynomial which
+    /// is the sum of `l` `d`-degree univariate polynomials where each coefficient
+    /// is sampled uniformly at random.
+    fn rand<R: Rng>(d: usize, num_vars: Option<usize>, rng: &mut R) -> Self;
 
-impl<'a, F: Field> Into<DensePolynomial<F>> for DenseOrSparsePolynomial<'a, F> {
-    fn into(self) -> DensePolynomial<F> {
-        match self {
-            DPolynomial(p) => p.into_owned(),
-            SPolynomial(p) => p.into_owned().into(),
-        }
-    }
+    /// Sample a random point from `Self::Domain`.  
+    fn rand_domain_point<R: Rng>(domain_size: Option<usize>, rng: &mut R) -> Self::Domain;
 }
 
-impl<'a, F: 'a + Field> TryInto<SparsePolynomial<F>> for DenseOrSparsePolynomial<'a, F> {
-    type Error = ();
+/// Describes the interface for univariate polynomials
+pub trait UVPolynomial<F: Field>: Polynomial<F> {
+    /// Constructs a new polynomial from a list of coefficients.
+    fn from_coefficients_slice(coeffs: &[F]) -> Self;
 
-    fn try_into(self) -> Result<SparsePolynomial<F>, ()> {
-        match self {
-            SPolynomial(p) => Ok(p.into_owned()),
-            _ => Err(()),
-        }
-    }
+    /// Constructs a new polynomial from a list of coefficients.
+    fn from_coefficients_vec(coeffs: Vec<F>) -> Self;
+
+    /// Returns the coefficients of `self`
+    fn coeffs(&self) -> &[F];
 }
 
-impl<'a, F: Field> DenseOrSparsePolynomial<'a, F> {
-    /// Checks if the given polynomial is zero.
-    pub fn is_zero(&self) -> bool {
-        match self {
-            SPolynomial(s) => s.is_zero(),
-            DPolynomial(d) => d.is_zero(),
-        }
-    }
+/// Describes the interface for univariate polynomials
+pub trait MVPolynomial<F: Field>: Polynomial<F> {
+    /// The type of the terms of `self`
+    type Term: multivariate::Term;
 
-    /// Return the degree of `self.
-    pub fn degree(&self) -> usize {
-        match self {
-            SPolynomial(s) => s.degree(),
-            DPolynomial(d) => d.degree(),
-        }
+    /// Constructs a new polynomial from a list of tuples of the form `(Self::Term, coeff)`
+    fn from_coefficients_slice(num_vars: usize, terms: &[(Self::Term, F)]) -> Self {
+        Self::from_coefficients_vec(num_vars, terms.to_vec())
     }
 
-    #[inline]
-    fn leading_coefficient(&self) -> Option<&F> {
-        match self {
-            SPolynomial(p) => p.last().map(|(_, c)| c),
-            DPolynomial(p) => p.last(),
-        }
-    }
+    /// Constructs a new polynomial from a list of tuples of the form `(Self::Term, coeff)`
+    fn from_coefficients_vec(num_vars: usize, terms: Vec<(Self::Term, F)>) -> Self;
```

Comment:
Does it make sense to move these methods into `Polynomial`? Then `UVPolynomial` will set `Term = F` or something along those lines, and `from_coefficients_vec` will just invoke that of `Polynomial`?

Reply:
Hmm so currently `UVPolynomial` doesn't have the `Term` associated type since I don't think it would ever be something besides `F`? But also the multivariate version of `from_coefficients_vec` additionally requires `num_vars` to be passed so I think it makes sense to have them separate considering that's what we did for `rand` as well.